File: cpp/medium/Trie.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;

class Trie {
public:
  Trie() : end(false), childs(26, nullptr) {}

  void insert(string word) {
    Trie *cur = this;
    for (int i = 0; i < word.size(); ++i) {
      vector<Trie *> &temp = cur->childs;
      char c = word[i];
      if (temp[c - 'a'] == nullptr) {
        Trie *newnode = new Trie();
        temp[c - 'a'] = newnode;
      }
      if (i == word.size() - 1)
        temp[c - 'a']->end = true;
      cur = temp[c - 'a'];
    }
  }

  bool search(string word) {
    Trie *travel = this;
    for (int i = 0; i < word.size(); ++i) {
      char c = word[i];
      if (travel && travel->childs[c - 'a']) {
        if (i == word.size() - 1 && travel->childs[c - 'a']->end == false)
          return false;
        travel = travel->childs[c - 'a'];
      } else {
        return false;
      }
    }
    return true;
  }

  bool startsWith(string prefix) {
    Trie *travel = this;
    for (int i = 0; i < prefix.size(); ++i) {
      char c = prefix[i];
      if (travel && travel->childs[c - 'a']) {
        travel = travel->childs[c - 'a'];
      } else {
        return false;
      }
    }
    return true;
  }

private:
  bool end;
  vector<Trie *> childs;
};

int main() {
  Trie *root = new Trie();
  root->insert("apple");
  cout << root->search("apple");
  return 0;
}

File: go/middle/Trie.go
package main

type Trie struct {
	val    byte
	childs [26]*Trie
	isEnd  bool
}

/** Initialize your data structure here. */
func Constructor() Trie {
	return Trie{val: ' '}
}

/** Inserts a word into the trie. */
func (this *Trie) Insert(word string) {
	for i := range word {
		if this.childs[word[i]-'a'] == nil {
			this.childs[word[i]-'a'] = &Trie{val: word[i]}
		}
		this = this.childs[word[i]-'a']
		if i == len(word)-1 {
			this.isEnd = true
		}
	}
}

/** Returns if the word is in the trie. */
func (this *Trie) Search(word string) bool {
	for i := range word {
		if this.childs[word[i]-'a'] == nil {
			return false
		}
		this = this.childs[word[i]-'a']
	}
	return this.isEnd
}

/** Returns if there is any word in the trie that starts with the given prefix. */
func (this *Trie) StartsWith(prefix string) bool {
	for i := range prefix {
		if this.childs[prefix[i]-'a'] == nil {
			return false
		}
		this = this.childs[prefix[i]-'a']
	}
	return true
}
